### PR TITLE
@skip, @include respect avoidOptionals

### PR DESCRIPTION
**Pull request by gilgardosh**
Related: https://github.com/dotansimha/graphql-code-generator/issues/5099

### CODE REVIEW

**Review by Cellule — COMMENTED** [review]
> Reviewed 10 of 10 files at r1, 42 of 42 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dotansimha/graphql-code-generator/5104)* status: all files reviewed, 2 unresolved discussions (waiting on @gilgardosh)

___
*[packages/plugins/typescript/operations/src/ts-selection-set-processor.ts, line 39 at r2](https://reviewable.io/reviews/dotansimha/graphql-code-generator/5104#-MMR4HPPD_dsTl0q13bu:-MMR4HPPD_dsTl0q13bv:b-ozp8ku) ([raw file](https://github.com/dotansimha/graphql-code-generator/blob/863409646cd79546fa6b634ea83ba89af542dd2b/packages/plugins/typescript/operations/src/ts-selection-set-processor.ts#L39)):*
> ```XML
> 
>     if (hasConditionals) {
>       if (
> ```

nit: 
```ts
const avoidOptional = (
  // TODO: check type and exec only if relevant
  this.config.avoidOptionals === true ||
  this.config.avoidOptionals.field ||
  this.config.avoidOptionals.inputValue ||
  this.config.avoidOptionals.object
)
const transform = avoidOptional ? "MakeMaybe" : "MakeOptional"
resString = `${
  this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : ''
}${transform}<${resString}, ${conditilnalsList.map(field => `'${field}'`).join(' | ')}>`;

```
___
*[packages/plugins/typescript/operations/tests/ts-documents.spec.ts, line 4770 at r2](https://reviewable.io/reviews/dotansimha/graphql-code-generator/5104#-MMR4kSG1-9gLQ6mXflB:-MMR4kSG1-9gLQ6mXflC:b-4x9ygd) ([raw file](https://github.com/dotansimha/graphql-code-generator/blob/863409646cd79546fa6b634ea83ba89af542dd2b/packages/plugins/typescript/operations/tests/ts-documents.spec.ts#L4770)):*
> ```XML
>     });
> 
>     it('On avoidOptionals:true, fileds with @skip, @include should make container resolve into MakeMaybe type', async () => {
> ```

nit: fields

**Comment by dotansimha** [comment]
@gilgardosh ready for a review? (btw, changeset is missing)

**Comment by gilgardosh** [comment]
@dotansimha Ready now

File: dev-test/githunt/typed-document-nodes.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';
 export type Maybe<T> = T | null;
 export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
-export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
-  { [SubKey in keyof Pick<T, K>]?: Maybe<Pick<T, K>[SubKey]> };
+export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
```
Comment (dotansimha) [inline dev-test/githunt/typed-document-nodes.ts:4]: I think we can experiment with TS generics inference here.
Something like :

```ts
export type MakeMaybe<T, K extends keyof T, AndOptional extends boolean> = AndOptional extends true ? Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> } : Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> ;
```

What do you think?

**Review by dotansimha — APPROVED** [review]

**Pull request merged**